---
title: "Differential expression among histology groups"
---

```{r echo=F, include=F}
library('tidyverse')
library('ComplexHeatmap')
```

### Data read and preparation
```{r include=F}
# Read clinical data and patient keys
key_fp = '../data/ClinicalLinkagewithFiles_niceNames.xlsx'
key_df = readxl::read_excel(key_fp, sheet=1, .name_repair='minimal') %>%
  janitor::clean_names() %>%
  filter(!reviewer_remove) %>% # Histologies suggested to be removed by reviewer
  mutate(rna_seq_mod=tolower(.[['rna_seq']]) %>%
           str_replace(., '\\-', '_')) %>%
  filter(disease_type == 'SAR - Sarcoma') %>%
  filter(rna_seq_mod != 'na')

rm(key_fp) # Clean environment
```

```{r include=F}
# Samples removed from clinical key data frame
key_df = key_df %>% 
  filter(rna_seq_mod != "sl424998") # No library prep info on meta data - Unable to apply Combat
```

```{r include=F}
# Read DTC immune clusters assignments (generated with MCPcounter immune scores)
# Refer to script "sarcoma_avatar_rnaseq_mcpdeconv.Rmd"
# Add data to clinical key
dtc_fp = '../data/immune_groups_deconvolution/mcpcounter_immune_clusters.csv'
key_df = read_delim(dtc_fp, delim=',', show_col_types=F, name_repair='minimal') %>%
  select(c('samplename', 'immunegroup')) %>%
  left_join(key_df, ., by=c('rna_seq_mod'='samplename'))

rm(dtc_fp) # Clean environment
```

```{r include=F}
# Recode sarcoma types with less than ten samples (many sarcoma types will be difficult to visualize)
sarcoma_abundant = names(table(key_df[['cdc_revision']]))[table(key_df[['cdc_revision']]) >= 5]
key_df = key_df %>%
  mutate(sarcoma_collapsed=case_when(cdc_revision %in% sarcoma_abundant ~ cdc_revision, TRUE ~ 'other')) %>%
  mutate(cdc_nice_name=case_when(sarcoma_collapsed == 'other' ~ 'Other', TRUE ~ cdc_nice_name))

rm(sarcoma_abundant) # Clean env
```

```{r include=F}
# Read gene expression summarized from isoforms (i.e., sum of isoforms for a given gene)
# Remove counts from the misc_RNA classes rRNAs, 'Y_RNA' and 'Metazoa_SRP'
# Remove pseudogenes, microRNAs, and snoRNAs
counts_sum_df = data.table::fread(file='../data/transformed_gene_expr_sum_isoforms.csv') %>%
  filter(!(gene_name %in% c('Y_RNA', 'Metazoa_SRP', '5_8S_rRNA', '5S_rRNA', '7SK', 'Vault'))) %>%
  filter(!str_detect(gene_name, "^SNORD|^SNORA|^MIR|^RNA5|^RNU|^U[0-9]+$")) %>% 
  select(c('gene_name', key_df[['rna_seq_mod']])) %>%
  column_to_rownames(var='gene_name')
```

```{r include=F}
# If column names in expression and sample names in meta data are not in the same order = ERROR
if(!identical(colnames(counts_sum_df), key_df[['rna_seq_mod']])){
  stop('Verify sample names are in the same order in expression data and metadata')
}
```

### DE gene tests
```{r include=F}
grps = list(
  leiomyo_grps = ifelse(key_df[['sarcoma_collapsed']] == 'leiomyosarcoma', 'leiomyosarcoma', 'other'),
  gist_grps = ifelse(key_df[['sarcoma_collapsed']] == 'gastrointestinal_stromal_tumor', 'gist', 'other'),
  myxlipo_grps = ifelse(key_df[['sarcoma_collapsed']] == 'myxoid_liposarcoma', 'myxoid_lipo', 'other'),
  ewing_grps = ifelse(key_df[['sarcoma_collapsed']] == 'ewing_sarcoma', 'ewing_sarcoma', 'other'),
  lipo_grps = ifelse(key_df[['sarcoma_collapsed']] %in% c('dedifferentiated_liposarcoma', 
                                                          'liposarcoma_NOS', 
                                                          'liposarcoma_well_differentiated'), 'lipo', 'other'))

# Find DE genes for certain histologies
top_de = list()
for(i in names(grps)){
  mtx = model.matrix(~0 + grps[[i]])
  colnames(mtx) = str_extract(colnames(mtx), paste0(unique(grps[[i]]), collapse='|'))
  
  modelfit = limma::lmFit(counts_sum_df, mtx)
  factors_tmp =  c(grep('other', colnames(coef(modelfit)), value=T, invert=T), 'other')
  contr = limma::makeContrasts(contrasts=paste0(factors_tmp, collapse=' - '), levels=colnames(coef(modelfit)))
  gene_fits = limma::contrasts.fit(modelfit, contr)
  ebayes_smooth = limma::eBayes(gene_fits)
  
  top_de[[i]] = limma::topTable(ebayes_smooth, sort.by="P", n=Inf)
  top_de[[i]] = top_de[[i]] %>% 
    rownames_to_column(var='gene') %>% arrange(adj.P.Val, desc(logFC))
  
  rm(mtx, modelfit, contr, gene_fits, ebayes_smooth) # Clean env
}

openxlsx::write.xlsx(top_de, '../results/differential_expression_histologies/diff_expr_genes_histologies.xlsx')
```

#### PLOTS OF GENES AMONG IMMUNEGROUPS
```{r include=F}
# Histology color palette
col_pal_hist = readRDS('../data/collapsed_histology_color-vec.rds')
names(col_pal_hist)[names(col_pal_hist) == 'other'] = 'Other'

# Create color palettes for primary/met and NMF categories
col_pal_met = as.vector(khroma::color('highcontrast')(2))
names(col_pal_met) = gsub('[-\\(\\)\\/]+', '', unique(key_df[['primary_met']]))

# Create color palettes for DTC clusters
col_pal_dtc = scales::hue_pal()(length(unique(key_df[['immunegroup']])))
names(col_pal_dtc) = c('A', 'B', 'C', 'D', 'E')
```

```{r include=F}
# Generate heatmap with subset of DE genes
# Order samples according to immunogroup assignments and histology
annots_df_ordered = key_df %>%
  select(c('rna_seq_mod', 'immunegroup', 'cdc_nice_name', 'primary_met')) %>%
  arrange(immunegroup, cdc_nice_name) %>%
  column_to_rownames('rna_seq_mod')

# Create heatmap annotation object
hm_ann = HeatmapAnnotation(df=annots_df_ordered %>%
                             select(c('cdc_nice_name', 'primary_met', 'immunegroup')),
                           col=list(cdc_nice_name=col_pal_hist,
                                    primary_met=col_pal_met,
                                    immunegroup=col_pal_dtc))

# Make color scale to accentuate the differences
hm_col = circlize::colorRamp2(c(-2.5, -1.5, 0, 1.5, 2.5), c("darkblue", "blue", "white", "red", "darkred"))
```

```{r include=F}
# Top DE genes by histology group
genes_keep = lapply(top_de, function(i){
  top_de_tmp = i %>% arrange(desc(logFC))
  genes_tmp = top_de_tmp[['gene']][ top_de_tmp[['adj.P.Val']] < 0.05 & top_de_tmp[['logFC']] > 1 ]
  genes_tmp = genes_tmp[1:20]
  return(genes_tmp)
})
genes_keep = unique(as.vector(unlist(genes_keep)))

# Subset genes and re-scale
counts_hm_mtx = counts_sum_df[na.omit(match(genes_keep, rownames(counts_sum_df))) , ]
counts_hm_mtx = t(scale(t(counts_hm_mtx)))

# Sort data according to MCPcounter
counts_hm_mtx = counts_hm_mtx[, match(rownames(annots_df_ordered %>% arrange(immunegroup, cdc_nice_name)), colnames(counts_hm_mtx))]

# Create and plot heatmap
gexpr_hm = Heatmap(counts_hm_mtx, 
             cluster_columns=T, cluster_rows=F, 
             show_row_names=T, show_column_names=F, 
             show_row_dend=F,
             row_title='Log-normalized\ngene expression',
             heatmap_legend_param=list(title="Scaled\ngene expr"),
             show_column_dend=F,
             row_names_side="left",
             column_split=annots_df_ordered[['mcpcounter_dtc']],
             bottom_annotation=hm_ann,
             col=hm_col)

pdf('../results/differential_expression_histologies/heatmap_subset_de_genes_histologies.pdf', height=14, width=8)
print(gexpr_hm)
dev.off()

rm(genes_keep, counts_hm_mtx, gexpr_hm) # Clean env
```

```{r include=F}
# Top DE genes by histology group
genes_keep = lapply(top_de, function(i){
  top_de_tmp = i %>% arrange(desc(logFC))
  genes_tmp = top_de_tmp[['gene']][ top_de_tmp[['adj.P.Val']] < 0.05 & top_de_tmp[['logFC']] > 1 ]
  genes_tmp = genes_tmp[1:5]
  return(genes_tmp)
})
genes_keep = unique(as.vector(unlist(genes_keep)))

# Subset genes and re-scale
counts_hm_mtx = counts_sum_df[na.omit(match(genes_keep, rownames(counts_sum_df))) , ]
counts_hm_mtx = t(scale(t(counts_hm_mtx)))

# Sort data according to MCPcluster
counts_hm_mtx = counts_hm_mtx[, match(rownames(annots_df_ordered %>% arrange(immunegroup, cdc_nice_name)), colnames(counts_hm_mtx))]

# Create and plot heatmap
gexpr_hm = Heatmap(counts_hm_mtx, 
             cluster_columns=T, cluster_rows=F, 
             show_row_names=T, show_column_names=F, 
             show_row_dend=F,
             row_title='Log-normalized\ngene expression',
             heatmap_legend_param=list(title="Scaled\ngene expr"),
             show_column_dend=F,
             row_names_side="left",
             column_split=annots_df_ordered[['mcpcounter_dtc']],
             bottom_annotation=hm_ann,
             col=hm_col)

pdf('../results/differential_expression_histologies/heatmap_subset_de_genes_histologies_small.pdf', height=10, width=8)
print(gexpr_hm)
dev.off()

rm(genes_keep, counts_hm_mtx, gexpr_hm) # Clean env
```

